lib/cosmos/core_ext/io.rb

```diff
@@ -20,74 +21,102 @@ class IO
+              current_timeout = timeout - total_timeout
+              total_timeout = timeout
+            else
+              # Or block forever
+              current_timeout = nil
+            end
           else
-            # Up our total time in select
-            total_timeout += current_timeout
-          end
-          if timeout and total_timeout > timeout
-            # Block for the remaining requested timeout
-            current_timeout = timeout - total_timeout
-            total_timeout = timeout
+            # Or it is bigger than the given timeout
+            if timeout and current_timeout >= timeout
+              # Block for the remaining requested timeout
+              current_timeout = timeout - total_timeout
+              total_timeout = timeout
+            else
+              # Up our total time in select
+              total_timeout += current_timeout
+            end
+            if timeout and total_timeout > timeout
+              # Block for the remaining requested timeout
+              current_timeout = timeout - total_timeout
+              total_timeout = timeout
+            end
           end
+          return nil if current_timeout and current_timeout < 0
         end
-        return nil if current_timeout and current_timeout < 0
-      end
-    end # while true
-  end # fast_select
+      end # while true
+    end # fast_select
 
-  # @param read_sockets [Array<IO>] IO objects to wait to be ready to read
-  # @param timeout [Numeric] Number of seconds to wait
-  def self.fast_read_select(read_sockets, timeout)
-    return fast_select(read_sockets, nil, nil, timeout)
+    # For non-windows systems, leave IO.select alone, but for windows
+    # monkey-patch IO.select to use our fast-select implementation.
+    if Kernel.is_windows?
+      def select(read_sockets = nil, write_sockets = nil, error_array = nil, timeout = nil)
+        return fast_select(read_sockets, write_sockets, error_array, timeout)
+      end
+    end
+    
+    # @param read_sockets [Array<IO>] IO objects to wait to be ready to read
+    # @param timeout [Numeric] Number of seconds to wait
+    def fast_read_select(read_sockets, timeout)
+      return fast_select(read_sockets, nil, nil, timeout)
+    end
+    
+    # @param write_sockets [Array<IO>] IO objects to wait to be ready to write
+    # @param timeout [Numeric] Number of seconds to wait
+    def fast_write_select(write_sockets, timeout)
+      return fast_select(nil, write_sockets, nil, timeout)
+    end
   end
 
-  # @param write_sockets [Array<IO>] IO objects to wait to be ready to write
-  # @param timeout [Numeric] Number of seconds to wait
-  def self.fast_write_select(write_sockets, timeout)
-    return fast_select(nil, write_sockets, nil, timeout)
+  # Alias the original close method
+  alias_method :__close__, :close
+
+  # Patch the close method so that it won't raise any exceptions
+  def close
+    begin
+      __close__
+    rescue
+    end
```

Comment:
We're ok with silently eating any exceptions raised by close? I guess there's not much more to do once you close the socket but I wanted to ask.

Reply:
Ryan and I talked about this; it probably isn't ideal, but those exceptions weren't helping much anyway-- typically they're only fired when we try to close something that is already closed.